clients/src/main/java/org/apache/kafka/clients/consumer/internals/StreamsGroupHeartbeatRequestManager.java

```diff
@@ -0,0 +1,387 @@
+                retryBackoffMaxMs,
+                jitter);
+            this.heartbeatIntervalMs = heartbeatIntervalMs;
+            this.heartbeatTimer = time.timer(heartbeatIntervalMs);
+        }
+
+        private void update(final long currentTimeMs) {
+            this.heartbeatTimer.update(currentTimeMs);
+        }
+
+        public void resetTimer() {
+            this.heartbeatTimer.reset(heartbeatIntervalMs);
+        }
+
+        @Override
+        public boolean canSendRequest(final long currentTimeMs) {
+            update(currentTimeMs);
+            return heartbeatTimer.isExpired() && super.canSendRequest(currentTimeMs);
+        }
+
+        private void updateHeartbeatIntervalMs(final long heartbeatIntervalMs) {
+            if (this.heartbeatIntervalMs == heartbeatIntervalMs) {
+                // no need to update the timer if the interval hasn't changed
+                return;
+            }
+            this.heartbeatIntervalMs = heartbeatIntervalMs;
+            this.heartbeatTimer.updateAndReset(heartbeatIntervalMs);
+        }
+    }
+
+    private final Logger logger;
+
+    private final int maxPollIntervalMs;
+
+    private final CoordinatorRequestManager coordinatorRequestManager;
+
+    private final HeartbeatRequestState heartbeatRequestState;
+
+    private final HeartbeatState heartbeatState;
+
+    private final StreamsMembershipManager membershipManager;
+
+    private final BackgroundEventHandler backgroundEventHandler;
+
+    private final HeartbeatMetricsManager metricsManager;
+
+    private StreamsRebalanceData streamsRebalanceData;
+
+    public StreamsGroupHeartbeatRequestManager(
+        final LogContext logContext,
+        final Time time,
+        final ConsumerConfig config,
+        final CoordinatorRequestManager coordinatorRequestManager,
+        final StreamsMembershipManager membershipManager,
+        final BackgroundEventHandler backgroundEventHandler,
+        final Metrics metrics,
+        final StreamsRebalanceData streamsRebalanceData
+    ) {
+        this.coordinatorRequestManager = coordinatorRequestManager;
+        this.logger = logContext.logger(getClass());
+        this.membershipManager = membershipManager;
+        this.backgroundEventHandler = backgroundEventHandler;
+        this.maxPollIntervalMs = config.getInt(CommonClientConfigs.MAX_POLL_INTERVAL_MS_CONFIG);
+        long retryBackoffMs = config.getLong(ConsumerConfig.RETRY_BACKOFF_MS_CONFIG);
+        long retryBackoffMaxMs = config.getLong(ConsumerConfig.RETRY_BACKOFF_MAX_MS_CONFIG);
+        this.heartbeatState = new HeartbeatState(streamsRebalanceData, membershipManager,
+            maxPollIntervalMs);
+        this.heartbeatRequestState = new HeartbeatRequestState(logContext, time, 0, retryBackoffMs,
+            retryBackoffMaxMs, maxPollIntervalMs);
+        this.metricsManager = new HeartbeatMetricsManager(metrics);
+        this.streamsRebalanceData = streamsRebalanceData;
```

Comment:
Maybe the order of assignment could agree with the order of parameters.

We could consider requiring non-null for many of these.